V2 API Support in HttpFilter

ramaraochavali opened the pull request:
Signed-off-by: Rama <ramaraochavali@gmail.com>

ramaraochavali commented on http-filter-example/envoy-v2.yaml:
```diff
@@ -0,0 +1,33 @@
+  listeners:
+  - name: listener_0
+    address:
+      socket_address: { address: {{ ip_loopback_address }}, port_value: 0 }
```
@htuch how do I refer this ip_loopback_address? json also seems to have the same way but not sure how it is resolved. Should I just give "127.0.0.1"?

htuch replied:
This is substituted by the integration test framework for you. Does it work?

ramaraochavali replied:
it did not work

ramaraochavali replied:
integration framework looks like expecting "envoy-v2.yaml.conf.json" - it is not doing it for the existing test as well. may be be i am doing some thing wrong

htuch replied:
Try `ntop_ip_loopback_address` here instead of `ip_loopback_address`. In general, `test/config/integration/server_xds*yaml` is a good reference for config fragments that can be consumed by the integration tests.

htuch replied:
I don't think you should need a YAML file with the new v2 style integration tests in any case. The core bootstrap proto is built programatically for you. Have you tried removing this file and the Bazel BUILD reference to it below?

ramaraochavali replied:
Yes. This is not required. I will nuke this, .json and v1 test case

htuch commented on http-filter-example/http_filter_v2_integration_test.cc:
```diff
@@ -0,0 +1,50 @@
+  void SetUp() override {
+    fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP1, version_));
+    registerPort("upstream_0", fake_upstreams_.back()->localAddress()->ip()->port());
+    createTestServer("http-filter-example/envoy-v2.yaml", {"http"});
```
For v2 testing, we have a somewhat better approach, where configs are built progamatically. See https://github.com/envoyproxy/envoy/blob/master/test/integration/ratelimit_integration_test.cc#L14 for example.

ramaraochavali replied:
I can adopt this. Not a big deal. I have made changes. But build is failing because it is not able to refer the external dep
 external_deps = [
        "envoy_filter_network_http_connection_manager"
    ]
It expects some changes in WORKSPACE - can you point to me to the required change in WORKSPACE?

ramaraochavali replied:
ok. looks like it has been moved from http to network scope. May be I am working with old - should i update envoy in this and test?

ramaraochavali commented:
@htuch I have adopted the v2 test process. I will have to update to latest envoy so that i can move http_connection_manager from "http" to "network" which I will do. In the mean time , can you please review the test and see if it makes sense?

Also when I run the test locally, it is failing with the following assertion - any idea?
[2017-12-02 14:09:29.759][13][warning][testing] external/envoy/test/test_common/environment.cc:82] Testing with IPv6 addresses may not be supported on this machine. If testing fails, set the environment variable ENVOY_IP_TEST_VERSIONS.
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from IpVersions/HttpFilterSamplev2IntegrationTest
[ RUN      ] IpVersions/HttpFilterSamplev2IntegrationTest.Test1/0
[2017-12-02 14:09:29.797][13][critical][assert] external/envoy/test/config/utility.cc:113] assert failure: port_idx == ports.size()

ramaraochavali commented:
@htuch I could by pass the error. Test seems to be running fine with ipv4. ipv6 could no test. Please check the code.

htuch commented on http-filter-example/BUILD:
```diff
@@ -1,11 +1,13 @@
-package(default_visibility = ["//visibility:public"])
+#package(default_visibility = ["//visibility:public"])
```
?

ramaraochavali replied:
just testing..will revert it

htuch commented on http-filter-example/BUILD:
```diff
@@ -46,3 +48,16 @@ envoy_cc_test(
+    srcs = ["http_filter_v2_integration_test.cc"],
+    data = [":envoy-v2.yaml"],
+    repository = "@envoy",
+    external_deps = ["envoy_filter_http_http_connection_manager"
```
Nit: join following line with this one.

htuch commented:
Does your environment run the IPv6 integration tests in the main Envoy repository?

ramaraochavali commented:
No. I am getting the following warning
Testing with IPv6 addresses may not be supported on this machine. If testing fails, set the environment variable ENVOY_IP_TEST_VERSIONS.

ramaraochavali commented:
@htuch Did the following changes
1.Bumped envoy to HEAD
2.Changed the http_connection_manger to refer to new "network" scope.
3.corrected build files.
Verified ipV6 and some warning coming on my machine and i think it is my local machine issue.
I think this should be good to go now. Please review once and let me know if any further changes are required.

htuch commented on http-filter-example/http_filter_v2_integration_test.cc:
```diff
@@ -0,0 +1,82 @@
+
+        });
+    named_ports_ = {"http"};
+    HttpIntegrationTest::initialize();
```
You call this twice now, once here and one in `SetUp`

ramaraochavali replied:
Yeah. so the basic setup should call initialize() and this initialize should call HttpIntegrationTest::initialize() similar to how RateLimitingTest does. I have changed that.

htuch commented on http-filter-example/http_filter_v2_integration_test.cc:
```diff
@@ -0,0 +1,82 @@
+#include "common/grpc/codec.h"
```
Nit: you can probably keep the old test name `http_filter_integration_test`. Either that, or update docs at https://github.com/envoyproxy/envoy-filter-example/blob/master/http-filter-example/README.md.

ramaraochavali replied:
@htuch given that v2 is going to be default choice, I changed the test name back to http_filter_integration_test rather than updating docs.

htuch commented on http-filter-example/http_filter_v2_integration_test.cc:
```diff
@@ -0,0 +1,82 @@
+      auto* sample_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      sample_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+      sample_cluster->set_name("service1");
+      sample_cluster->mutable_http2_protocol_options();
```
Why do you setup an additional service here? The basic config already has a cluster https://github.com/envoyproxy/envoy/blob/master/test/config/utility.cc#L41.

ramaraochavali replied:
Initially I did not setup. But with out that it fails with assertion at https://github.com/envoyproxy/envoy/blob/master/test/config/utility.cc#L110. That is probably because of the way the for loop is setup at 
https://github.com/envoyproxy/envoy/blob/master/test/config/utility.cc#L100??. So I have added this.

htuch commented on http-filter-example/http_filter_v2_integration_test.cc:
```diff
@@ -0,0 +1,82 @@
+          hcm.mutable_route_config()
+              ->mutable_virtual_hosts(0)
+              ->mutable_routes(0)
+              ->mutable_route();
```
Ditto.

ramaraochavali replied:
Yes. This is not required. Removed.

ramaraochavali commented:
@htuch I have addressed all the comments. Should be go to go now.

htuch commented on http-filter-example/http_filter_integration_test.cc:
```diff
@@ -10,17 +14,35 @@ class HttpFilterSampleIntegrationTest : public HttpIntegrationTest,
+        "{ name: sample, config: {} }");
+
+    config_helper_.addConfigModifier([](envoy::api::v2::Bootstrap &bootstrap) {
+      auto *sample_cluster = bootstrap.mutable_static_resources()->add_clusters();
```
I think it probably would have been better to point at test/integration/integration_test.h` rather than the HTTP2 test, sorry about that. You can remove this `addConfigModifier` if you also remove the `createUpstreams`, which is only needed for H2 backends. We only care about simple H1 in this example.

htuch commented on http-filter-example/http_filter_integration_test.cc:
```diff
@@ -10,17 +14,35 @@ class HttpFilterSampleIntegrationTest : public HttpIntegrationTest,
+
+
+  void cleanup() {
+    if (fake_upstream_connection_ != nullptr) {
```
This can probably also go.

ramaraochavali commented:
@htuch addressed the comments. Please check once.

htuch commented:
Thanks @ramaraochavali!
